Secret rotator gsm functions

b01901143 opened the pull request:
First push for secret rotator
Imlementing all secret manager related functions.
Leaving provisioners as TODO.

b01901143 committed:
Secret rotator gsm functions

First push for secret rotator
Imlementing all secret manager related functions.
Leaving provisioners as TODO.

b01901143 commented:
This testings will be included in next PR, because it would be of size XXL altogether.

/assign @spiffxp

b01901143 committed:
Add tests for secret rotator.

Imlementing all secret manager related functions.
Testings for `rotator.Refresh()` and `rotator.Deactivate()`
Leaving provisioners as TODO.

spiffxp commented on secret-rotator/client/client.go:
```diff
@@ -0,0 +1,289 @@
+	name := "projects/" + project + "/secrets/" + id + "/versions/" + version
+
+	switch state {
+	case 1: // SecretVersion_ENABLED
```
Do these live as constants somewhere?

spiffxp replied:
This feels like an unfriendly API, I guess I'm curious why we didn't make similar `EnableSecretVersion` etc. methods

b01901143 replied:
Yes, it's defined in https://github.com/googleapis/go-genproto/blob/da3ae01ef02dd0f1909d902122c432a450c6efb3/googleapis/cloud/secretmanager/v1/resources.pb.go#L46

b01901143 replied:
I was imported as alias `secretmanagerpb`, and I thought it's too lengthy to write `secretmanagerpb.SecretVersion_ENABLED` everywhere.

But thiking again now, maybe I could just make the alias shorter and use something like `gsmpb.SecretVersion_ENABLED` .

b01901143 replied:
> This feels like an unfriendly API, I guess I'm curious why we didn't make similar `EnableSecretVersion` etc. methods

Because I was trying to avoid the function naming of `client.EnableSecretVersion()`.
I have the `secretmanager.Client` embedded in `rotator.Client`, like 
```
type Client struct {
	*secretmanager.Client
}
``` 
so previously it always gave me errors if I had identical function names with `secretmanager.Client`.

But this could actually be easily solved, just call `cl.Client.EnableSecretVersion` within a `rotator.Client` method.
So I think I'll break this into `EnableSecretVersion`, `DisableSecretVersion`, and `DestroySecretVersion`.

spiffxp commented on secret-rotator/config/rotated-secret-config.go:
```diff
@@ -0,0 +1,93 @@
+	"time"
+)
+
+// Structs for rotated secret configuration
```
These need comments explaining what they each are

b01901143 replied:
Added.

spiffxp commented on secret-rotator/config/rotated-secret-config.go:
```diff
@@ -0,0 +1,93 @@
+type RotatedSecretType struct {
+	ServiceAccountKey *svckey.ServiceAccountKeySpec `yaml:"serviceAccountKey,omitempty"`
+}
+type RefreshStrategy struct {
+	Interval time.Duration `yaml:"interval,omitempty"`
+	Cron     string        `yaml:"cron,omitempty"`
+}
```
As an example of why these need comments: does this mean a strategy should include both an interval AND a cron? Or either?

b01901143 replied:
Adding comments to these structs

spiffxp commented on secret-rotator/config/rotated-secret-config.go:
```diff
@@ -0,0 +1,93 @@
+		return secretType.ServiceAccountKey.Type()
+	} else {
+		// TODO: other types of secrets
+		return ""
```
you should use a constant for this, and probably something like "UNKNOWN" so it's not confused with an default-initialized string

b01901143 replied:
Set to "UNKNOWN" for unknown types.

spiffxp commented on secret-rotator/config/rotated-secret-config.go:
```diff
@@ -0,0 +1,93 @@
+	if len(config.Specs) == 0 {
+		return fmt.Errorf("Empty secret sync configuration.")
+	}
+	// TODO: validate Project, Secret, Type, Refresh, GracePeriod fields
```
I would like to see this

b01901143 replied:
Added.

spiffxp replied:
for followup: you can drop this TODO now that it's been implemented

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,172 @@
+// RunOnce checks all rotated secrets in Agent.Config().Specs
+// Pops error message for any failure in refreshing or deactivating each secret.
+func (r *SecretRotator) RunOnce() {
+	// iterate on copy of Specs instead of index,
+	// so that the update in Agent.config will only be observed outside of the loop
```
Not sure I follow what this comment means?

b01901143 replied:
I was trying to express the difference between https://play.golang.org/p/Tq8SllIUJN6
and https://play.golang.org/p/2ULQBc16U-V,
if the update happens in the middle of a run of the for loop.

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,172 @@
+// If it does, Refresh provisions a new secret and updates the Secret Manager secret.
+// Returns true if the secret is refreshed.
+func (r *SecretRotator) Refresh(rotatedSecret config.RotatedSecretSpec, now time.Time) (bool, error) {
+	createTime, err := r.Client.GetCreateTime(rotatedSecret.Project, rotatedSecret.Secret, "latest")
+	if err != nil {
+		return false, err
+	}
```
From one perspective, it's intuitive that this won't bother to rotate non-existent secrets.  However this means I as a human have to provision the secret myself the first time.  Hopefully I did it the same way the rotator's going to do it next time, but I'm going to have to wait `interval` to find out.

From another perspective, we use "upsert" so provisioning would be the same if there was a secret there or not.  If the rotator created the initial secret version, I could find out right away if it's creating secrets I can use.  This seems more user friendly?

b01901143 replied:
Yeah I think we can implement an "upsert" here. If the Secret manager doesn't yet exist or is empty (with no version), the rotator simply create the first version and attach the "v1: key1" label to its metatdata.

This also make me think, it seems like the rotator should also validate that the secret versions align with the labels. Maybe this should be done every time before `rotator.Refresh()`?

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,172 @@
+				return false, err
+			}
+
+			err = r.Client.UpsertSecretLabel(rotatedSecret.Project, rotatedSecret.Secret, "v"+latestVersion, newId)
```
I feel like the "v" needs some explanation

b01901143 replied:
Added explanation in the comments

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,172 @@
+		// the refresh stratetgy is refreshInterval
+		// check the elapsed time from its createTime to now.
+		if now.After(createTime.Add(rotatedSecret.Refresh.Interval)) {
+			// provision a new secret
```
I would see if you can avoid nesting the provisioning block of code in the if checking whether we should provision
e.g.
```golang
if !shouldRefresh() {
  return false, nil
}
doTheProvisioning()
```

b01901143 replied:
Separate this into `rotator.ShouldRefresh(config, now)` and `rotator.Refresh(config, now)`.
`rotator.ShouldRefresh(config, now)` is used to decide whether a secret needs to be refreshed.

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,172 @@
+		v, _ := strconv.Atoi(curVersion)
+		nextVersion := strconv.Itoa(v + 1)
+
+		// check if curVersion exists
+		err = r.Client.ValidateSecretVersion(rotatedSecret.Project, rotatedSecret.Secret, curVersion)
+		if err != nil {
+			return err
+		}
```
We probably want to keep checking the rest of the labels anyway?

b01901143 replied:
Instead of returning error, now logging error message and continue.

b01901143 replied:
Also broke this into `rotator.ShouldDeactivate(secret, version, now)` and `rotator.Deactivate(secret, now)`

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,172 @@
+		}
+
+		if now.After(nextCreateTime.Add(rotatedSecret.GracePeriod)) {
+			// deactivate the old secret
+			err := r.Provisioners[rotatedSecret.Type.Type()].Deactivate(labels, curVersion)
+			if err != nil {
+				return err
+			}
+
+			// TODO: disable or destroy Secret Manager secret version curVersion?
+			// 1: SecretVersion_ENABLED, 2: SecretVersion_DISABLED, 3: SecretVersion_DESTROYED
+			err = r.Client.ChangeSecretVersionState(rotatedSecret.Project, rotatedSecret.Secret, curVersion, 2)
+			if err != nil {
+				return err
+			}
```
You could also consider a graceful deactivation path:
- first time through, active versions are disabled
- next time through, disabled versions are destroyed (and the corresponding secret deactivated)

This allows a nice grace period if there's human error involved in generating the new secret, giving us the opportunity to go back to using the old secret.

If you don't want to allow for such an option, then I don't think there's much point in keeping a secret version around as disabled if it's been deactivated.

b01901143 replied:
After discussing, if the service account key itself is deleted, there's no meaning to keep the copy of that private key.
We'll just destroy a version when it went out of its grace period.

spiffxp commented on secret-rotator/svckey/svckey.go:
```diff
@@ -0,0 +1,58 @@
+// returns the key-id and private-key data of the created key if successful,
+// otherwise returns error
+func (p *Provisioner) CreateNew(labels map[string]string) (string, []byte, error) {
+	// TODO: provision new service account key
```
would like to see this

b01901143 replied:
Implemented, will include in next commit.

spiffxp commented on secret-rotator/svckey/svckey.go:
```diff
@@ -0,0 +1,58 @@
+// Deactivate deletes an existing service account key specified by labels and version,
+// returns nil if successful, otherwise error
+func (p *Provisioner) Deactivate(labels map[string]string, version string) error {
+	// TODO: delete old service account key
```
same here

b01901143 replied:
Implemented, will include in next commit.

spiffxp commented on secret-rotator/tests/mockclient.go:
```diff
@@ -0,0 +1,213 @@
+)
+
+type MockClient struct {
+	Secrets map[string]map[string]*Secret
```
What is this a map of? Guessing project -> secret name -> secret?

b01901143 replied:
correct. Will add comments to these.

spiffxp commented on secret-rotator/tests/mockclient.go:
```diff
@@ -0,0 +1,213 @@
+	version := strconv.Itoa(len(cl.Secrets[project][id].Versions) + 1)
+	cl.Secrets[project][id].Versions[version] = &Version{
+		Data:  data,
+		State: 1, // SecretVersion_ENABLED
```
Really feels like these should be constants somewhere.  If upstream isn't providing them, then you should define them someplace

b01901143 replied:
Switched to using `secretmanagerpb.SecretVersion_State`

spiffxp commented on secret-rotator/tests/mockprovisoner.go:
```diff
@@ -0,0 +1,48 @@
+	// TODO: provision new service account key
+	name := fmt.Sprintf("projects/%s/serviceAccounts/%s", labels["project"], labels["service-account"])
+	klog.V(2).Infof("Provision a new secret of %s", name)
+	return p.NewSecretID, p.NewSecretValue, nil
```
I wonder if it would be better to provision random strings instead of fixed strings?

b01901143 replied:
I found it hard to unit test if I provision random strings.

b01901143 replied:
It gets a little messier to test after switching to provision random strings.

spiffxp commented on secret-rotator/rotator/rotator_test.go:
```diff
@@ -0,0 +1,425 @@
+								},
+							},
+							Labels: map[string]string{
+								"project":         "project-1",
+								"service-account": "service-foo",
```
Thinking about this more, maybe these don't make sense to store as label.  You don't have test cases or code written yet to decide how to handle the case where the rotatedsecret's spec for these differs from what's already stored in the secret.  I guess I would expect them to always be updated to whatever the rotated secret has in its spec, at which point I wonder why these are here at all.  It's informative for a human manually looking at labels on this secret, but maybe not necessary?

b01901143 replied:
discussed offline. will take the config as reference, but also update the Secret Manager labels within each run.

spiffxp reviewed (commented):
This looks mostly straightforward but I feel like without at least the TODO for service account key provisioning this is incomplete

b01901143 committed:
Add Service Account Key Provisioner

Implement provisioner for service account keys.
Support upserting a new rotated secret to an existing project.
Store the base64 decoded private key data into Secret Manager.

b01901143 committed:
Modify Makefile

binaries are built into the `_output/bin/` folder.
modify Dockerfile accordingly.
include secret-rotator in images/cloudbuild.yaml

spiffxp commented on secret-rotator/config/rotated-secret-config.go:
```diff
@@ -0,0 +1,135 @@
+		}
+
+		// validate there's only one secret type
+		// TODO
```
for followup: what's this TODO mean?

b01901143 replied:
Just to remind me to implement the validation that there's only one secret type, after provisioners for other secret types are implemented in the future.
I'll add more detail to this comment

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,263 @@
+// RunOnce checks all rotated secrets in Agent.Config().Specs
+// Pops error message for any failure in refreshing or deactivating each secret.
+func (r *SecretRotator) RunOnce() {
+	// iterating on rotatedSecret instead of index so that the config stays consistent within each iteration,
+	// even if a config update occurs in the middle of the loop.
+	for _, rotatedSecret := range r.Agent.Config().Specs {
```
For followup: I don't think avoiding the index saves you from this

b01901143 replied:
I tried to simulate this kind of situation where the update happens in the middle of the loop:
https://play.golang.org/p/Tq8SllIUJN6

and it iterates through the old config as what we would desire.

I think this indicates it does?

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,263 @@
+func (r *SecretRotator) ShouldRefresh(rotatedSecret config.RotatedSecretSpec, now time.Time) (bool, error) {
+	err := r.Client.ValidateSecretVersion(rotatedSecret.Project, rotatedSecret.Secret, "1")
+	if err != nil {
+		// create the secret and/or dirst version if it does not already exist
```
```suggestion
		// create the secret and/or first version if it does not already exist
```

spiffxp commented on secret-rotator/rotator/rotator.go:
```diff
@@ -0,0 +1,263 @@
+			return true, nil
+		}
+	} else {
+		// the refresh strategy is cron
+		// TODO
```
For followup: I would consider cron an invalid strategy in `Validate` if it's still unimplemented.  Consider maybe whether you'd like to drop cron entirely?

b01901143 replied:
I think I will drop cron entirely for now, but keeping the current `RotatedSecretConfig` structure so that there's still room to support cron in the future.

spiffxp commented on secret-rotator/svckey/svckey.go:
```diff
@@ -0,0 +1,103 @@
+// CreateNew provisions a new service account key,
+// returns the key-id and private-key data of the created key if successful,
+// otherwise returns error
+func (p *Provisioner) CreateNew(labels map[string]string) (string, []byte, error) {
+	name := fmt.Sprintf("projects/%s/serviceAccounts/%s@%s.iam.gserviceaccount.com", labels["project"], labels["service-account"], labels["project"])
```
For followup: what happens if these labels don't exist?  Maybe this is a sign `map[string]string` isn't the right interface into this function.  What would the drawbacks be to using `project, serviceAccount string`?

b01901143 replied:
I think the drawbacks are that for other types of provisioner (e.g. ssh key provisioner), `project, serviceAccount string` wouldn't be the right things to pass in, and might also require different number of fields.
I was trying to come up with an mostly general `Provisioner` interface. Although we can use variadic function parameters for `Provisioner.CreateNew(para ...string)`, that still implicitly requires those labels to exist.

Answering the first question, these labels should exist since they are attach to the input `labels` by https://github.com/kubernetes-sigs/k8s-gsm-tools/blob/15c29995e3dedf5e1f15e3667f65c7871204d7f3/secret-rotator/svckey/svckey.go#L42 and https://github.com/kubernetes-sigs/k8s-gsm-tools/blob/15c29995e3dedf5e1f15e3667f65c7871204d7f3/secret-rotator/rotator/rotator.go#L104.

There's one thing missing though is the Validate function for `ServiceAccountKeySpec` to ensure `Project` and `ServiceAccount` exist in the configMap.

spiffxp commented on secret-rotator/svckey/svckey.go:
```diff
@@ -0,0 +1,103 @@
+
+// Deactivate deletes an existing service account key specified by labels and version,
+// returns nil if successful, otherwise error
+func (p *Provisioner) Deactivate(labels map[string]string, version string) error {
+	// keys in format of "v%d" indicate that they are (version: id) pairs attached by the rotator
+	// the reason for the prefix "v" is that Secret Manager labels need to begin with a lowwer case letter
+	name := fmt.Sprintf("projects/%s/serviceAccounts/%s@%s.iam.gserviceaccount.com/keys/%s", labels["project"], labels["service-account"], labels["project"], labels["v"+version])
```
Same question here except `project, serviceAccount, keyId string`

spiffxp commented on secret-rotator/tests/mockclient.go:
```diff
@@ -0,0 +1,255 @@
+// GetSecretVersionData gets the data of the secret version specified by project, id, version.
+// Returns secret value if successful, otherwise error
+func (cl *MockClient) GetSecretVersionData(project, id, version string) ([]byte, error) {
+	err := cl.ValidateSecretVersion(project, id, version)
+	if err != nil {
+		return nil, err
+	}
+
+	if version == "latest" {
+		version = strconv.Itoa(len(cl.Secrets[project][id].Versions))
+	}
```
For followup: this pattern seems reused a lot in this file.  Is there some way you could reuse this and avoid the copy-paste?

b01901143 replied:
Added an utility function `ValidateAndConvertVersion()` that checks if the secret version exists and also converts the version into the string of version number if version is "latest".

spiffxp reviewed (approved):
/approve
/lgtm
Suggestions for a followup PR
